wallets/blink/client.js

```diff
@@ -0,0 +1,181 @@
+      await new Promise(resolve => setTimeout(resolve, 100))
+
+      const txInfo = await getTxInfo(authToken, wallet, invoice)
+      // settled
+      if (txInfo.status === 'SUCCESS') {
+        if (!txInfo.preImage) throw new Error('no preimage')
+        return txInfo.preImage
+      }
+      // failed
+      if (txInfo.status === 'FAILED') {
+        throw new Error(txInfo.error || 'failed to pay invoice')
+      }
+      // still pending
+      // retry later
+    }
+  }
+
+  // this should never happen
+  throw new Error('unexpected error')
+}
+
+async function getTxInfo (authToken, wallet, invoice) {
+  const walletId = wallet.id
+  let out
+  try {
+    out = await request(authToken, `
+      query GetTxInfo($walletId: WalletId!, $paymentRequest: LnPaymentRequest!) {
+        me {
+          defaultAccount {
+            walletById(walletId: $walletId) {
+              transactionsByPaymentRequest(paymentRequest: $paymentRequest) {
+                status
+                direction
+                settlementVia {
+                    ... on SettlementViaIntraLedger {
+                        preImage
+                    }
+                    ... on SettlementViaLn {
+                        preImage
+                    }
+                }
+              }
+            }
+          }
+        }
+      }
+    `,
+    {
+      paymentRequest: invoice,
+      walletId
+    })
+  } catch (e) {
+    // something went wrong during the query,
+    // maybe the connection was lost, so we just return
+    // a pending status, the caller can retry later
+    return {
+      status: 'PENDING',
+      preImage: null,
+      error: ''
+    }
+  }
+  const tx = out.data.me.defaultAccount.walletById.transactionsByPaymentRequest.find(t => t.direction === 'SEND')
+  if (!tx) {
+    // the transaction was not found, something went wrong
+    return {
+      status: 'FAILED',
+      preImage: null,
+      error: 'transaction not found'
+    }
+  }
+  const status = tx.status
+  const preImage = tx.settlementVia.preImage
+  return {
+    status,
+    preImage,
+    error: ''
+  }
+}
+
+async function getWallet (authToken, currency = 'BTC') {
+  const out = await request(authToken, `
+    query me {
+        me {
+            defaultAccount {
+                wallets {
+                    id
+                    walletCurrency
+                }
+            }
+        }
+    }
+  `, {})
+  const wallets = out.data.me.defaultAccount.wallets
+  for (const wallet of wallets) {
+    if (wallet.walletCurrency === currency) {
+      return wallet
+    }
+  }
+  throw new Error(`wallet ${currency} not found`)
+}
+
+async function request (authToken, query, variables = {}) {
+  const options = {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json',
+      'X-API-KEY': authToken
+    },
+    body: JSON.stringify({ query, variables })
+  }
+  return fetch(galoyBlinkUrl, options).then(response => response.json())
```

Comment:
If I enter an invalid blink API key, I get this error:

```
0s	[blink]	ERROR	Unexpected token '<', "<html> <h"... is not valid JSON
4s	[blink]	INFO	trying to fetch wallet
```

This is not a user-friendly error message; it looks like a bug. Fortunately, it does respond with `401 Unauthorized` (and not `200 OK` like most GraphQL APIs even on errors) so we can use it:

```suggestion
  const res = await fetch(galoyBlinkUrl, options)
  if (res.status === 401) {
    throw new Error('API responded with 401 Unauthorized')
  }
  return await res.json()
```

Reply:
done, i've changed it to catch other errors as well